native/libp2p_port/internal/discovery/discovery.go

```diff
@@ -0,0 +1,103 @@
+package discovery
+
+import (
+	"crypto/rand"
+	"fmt"
+	"libp2p_port/internal/port"
+	"libp2p_port/internal/proto_helpers"
+	"libp2p_port/internal/reqresp"
+	"libp2p_port/internal/utils"
+	"net"
+
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/p2p/enr"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+type Discoverer struct {
+	port           *port.Port
+	discv5_service *discover.UDPv5
+}
+
+func NewDiscoverer(p *port.Port, listener *reqresp.Listener, config *proto_helpers.Config) Discoverer {
+	if !config.UseDiscv5 {
+		return Discoverer{}
+	}
+	udpAddr, err := net.ResolveUDPAddr("udp", config.DiscoveryAddr)
+	utils.PanicIfError(err)
+	conn, err := net.ListenUDP("udp", udpAddr)
+	utils.PanicIfError(err)
+	intPrivKey, _, err := crypto.GenerateSecp256k1Key(rand.Reader)
+	utils.PanicIfError(err)
+	privKey, err := utils.ConvertFromInterfacePrivKey(intPrivKey)
+	utils.PanicIfError(err)
+
+	bootnodes := make([]*enode.Node, 0, len(config.Bootnodes))
+
+	for _, strBootnode := range config.Bootnodes {
+		bootnode, err := enode.Parse(enode.ValidSchemes, strBootnode)
+		utils.PanicIfError(err)
+		bootnodes = append(bootnodes, bootnode)
+	}
+
+	cfg := discover.Config{
+		// These settings are required and configure the UDP listener:
+		PrivateKey: privKey,
+
+		// These settings are optional:
+		// NetRestrict *netutil.Netlist  // list of allowed IP networks
+		Bootnodes: bootnodes, // list of bootstrap nodes
+		// Unhandled   chan<- ReadPacket // unhandled packets are sent on this channel
+		// Log         log.Logger        // if set, log messages go here
+
+		// V5ProtocolID configures the discv5 protocol identifier.
+		// V5ProtocolID *[6]byte
+
+		// ValidSchemes enr.IdentityScheme // allowed identity schemes
+		// Clock        mclock.Clock
```

Comment:
We should move the optional parameters comment outisde of this assignment, it makes it a bit hard to read. Is this a note? A TODO?